core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala

```diff
@@ -17,18 +17,137 @@
 
 package org.apache.spark.ui.jobs
 
-import scala.xml.{Node, NodeSeq}
+import scala.xml.{Node, NodeSeq, Unparsed}
 
+import java.util.Date
 import javax.servlet.http.HttpServletRequest
 
-import org.apache.spark.ui.{WebUIPage, UIUtils}
+import org.apache.spark.ui.{UIUtils, WebUIPage}
 import org.apache.spark.ui.jobs.UIData.JobUIData
+import org.apache.spark.JobExecutionStatus
 
 /** Page showing list of all ongoing and recently finished jobs */
 private[ui] class AllJobsPage(parent: JobsTab) extends WebUIPage("") {
   private val startTime: Option[Long] = parent.sc.map(_.startTime)
   private val listener = parent.listener
 
+  private def applicationTimelineView(jobs: Seq[JobUIData], now: Long): Seq[Node] = {
+    val jobEventJsonAsStrSeq = jobs.flatMap { jobUIData =>
+      val jobId = jobUIData.jobId
+      val status = jobUIData.status
+      val submissionTimeOpt = jobUIData.submissionTime
+      val completionTimeOpt = jobUIData.completionTime
+
+      if (status == JobExecutionStatus.UNKNOWN || submissionTimeOpt.isEmpty ||
+        completionTimeOpt.isEmpty && status != JobExecutionStatus.RUNNING) {
+        None
+      }
+
+      val submissionTime = submissionTimeOpt.get
+      val completionTime = completionTimeOpt.getOrElse(now)
+      val classNameByStatus = status match {
+        case JobExecutionStatus.SUCCEEDED => "succeeded"
+        case JobExecutionStatus.FAILED => "failed"
+        case JobExecutionStatus.RUNNING => "running"
+      }
+
+      val jobEventJsonAsStr =
+        s"""
+           |{
+           |  'className': 'job application-timeline-object ${classNameByStatus}',
```

Comment:
it might be nice to factor these out into a utility function:

```
def generateTimelineEventJSON(
  className: String,
  group: String,
  start: Date,
  end: Option[Date],
  title: String)
```